File: tools/glosh-remote-spike/app/src/test/java/com/glosh/remote/spike/guide/autopilot/AdaptiveAutopilotPlannerTest.java
package com.glosh.remote.spike.guide.autopilot;

import static org.junit.Assert.assertEquals;

import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Action;
import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Candidate;
import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Confidence;
import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Observation;
import com.glosh.remote.spike.guide.autopilot.AutopilotContract.Screen;
import com.glosh.remote.spike.guide.autopilot.AutopilotContract.SnapshotAuthority;

import org.junit.Test;

import java.util.List;

public class AdaptiveAutopilotPlannerTest {
    private static final SnapshotAuthority SAFE = new SnapshotAuthority(
            true, 2, true, true, true, false);
    private final AdaptiveAutopilotPlanner planner = new AdaptiveAutopilotPlanner();

    @Test
    public void referencePlannerCoverage() {
        int checks = 0;
        checks += expect(Action.DONE, base().supportConnected(true).build());
        checks += expect(Action.CONNECT_SUPPORT, base().adbConnected(true).build());
        checks += expect(Action.UNSUPPORTED_ANDROID, base().androidApi(29).build());
        checks += expect(Action.ASK_ENABLE_ACCESSIBILITY, base().accessibilityEnabled(false).build());
        checks += expect(Action.ASK_ALLOW_RESTRICTED_SETTINGS,
                base().accessibilityEnabled(false).restrictedSettingsRequired(true).build());
        checks += expect(Action.ASK_CONNECT_WIFI, base().wifiReady(false).build());
        checks += expect(Action.POLICY_BLOCKED, base().wirelessPolicyBlocked(true).build());
        checks += expect(Action.TRY_ADB_RECONNECT,
                base().previousPairingKnown(true).reconnectAttempted(false).build());
        checks += expect(Action.OPEN_DEVELOPER_SETTINGS,
                base().previousPairingKnown(true).reconnectAttempted(true).screen(Screen.APP).build());
        checks += expect(Action.OPEN_DEVELOPER_SETTINGS, base().screen(Screen.APP).build());
        checks += expect(Action.OPEN_DEVICE_INFO_SETTINGS,
                base().screen(Screen.SETTINGS_HOME).directDevProbeAttempted(true).build());
        checks += expect(Action.FALLBACK_GUIDE,
                base().oem("Xiaomi").screen(Screen.SETTINGS_HOME).directDevProbeAttempted(true).build());
        checks += expect(Action.CLICK_SOFTWARE_INFO,
                base().screen(Screen.ABOUT_PHONE).candidate(candidate("software_info")).build());
        for (int tap = 0; tap < 7; tap++) {
            checks += expect(Action.CLICK_BUILD_NUMBER,
                    base().screen(Screen.SOFTWARE_INFO).candidate(candidate("build_number"))
                            .buildTapsDone(tap).build());
        }
        checks += expect(Action.OPEN_DEVELOPER_SETTINGS,
                base().screen(Screen.SOFTWARE_INFO).candidate(candidate("build_number"))
                        .buildTapsDone(7).build());
        checks += expect(Action.WAIT_USER_CREDENTIAL, base().screen(Screen.CREDENTIAL_PROMPT).build());
        checks += expect(Action.CLICK_WIRELESS_DEBUGGING,
                base().screen(Screen.DEVELOPER_OPTIONS).candidate(candidate("wireless_debugging")).build());
        checks += expect(Action.ENABLE_WIRELESS_DEBUGGING,
                base().screen(Screen.WIRELESS_DEBUGGING)
                        .candidate(candidate("wireless_debugging_toggle")).wirelessEnabled(false).build());
        checks += expect(Action.CLICK_PAIR_WITH_CODE,
                base().screen(Screen.WIRELESS_DEBUGGING)
                        .candidate(candidate("pair_with_code")).wirelessEnabled(true).build());
        checks += expect(Action.WAIT_STABLE,
                base().screen(Screen.WIRELESS_DEBUGGING).candidate(candidate("pair_with_code")).build());
        checks += expect(Action.ACCEPT_NETWORK_CONFIRMATION,
                base().screen(Screen.NETWORK_CONFIRMATION)
                        .candidate(candidate("network_confirm_positive")).build());
        checks += expect(Action.AUTO_PAIR_WITH_CODE,
                base().screen(Screen.PAIRING_DIALOG).pairCodeCandidates(List.of("123456"))
                        .pairingContextHigh(true).build());
        checks += expect(Action.SHOW_MANUAL_PAIR_CODE,
                base().screen(Screen.PAIRING_DIALOG).pairCodeCandidates(List.of("123456", "654321"))
                        .pairingContextHigh(true).build());
        checks += expect(Action.SHOW_MANUAL_PAIR_CODE,
                base().screen(Screen.PAIRING_DIALOG).pairCodeCandidates(List.of("123456"))
                        .pairingContextHigh(false).build());
        checks += expect(Action.SHOW_MANUAL_PAIR_CODE,
                base().screen(Screen.PAIRING_DIALOG).pairCodeCandidates(List.of("123456"))
                        .pairingContextHigh(true).requestActive(false).build());
        checks += expect(Action.WAIT_STABLE,
                base().screen(Screen.PAIRING_DIALOG)
                        .authority(new SnapshotAuthority(true, 1, true, true, true, false))
                        .pairCodeCandidates(List.of("123456")).pairingContextHigh(true).build());
        checks += unsafe(new SnapshotAuthority(true, 1, true, true, true, false), candidate("software_info"));
        checks += unsafe(new SnapshotAuthority(false, 2, true, true, true, false), candidate("software_info"));
        checks += unsafe(new SnapshotAuthority(true, 2, true, true, true, true), candidate("software_info"));
        checks += unsafe(new SnapshotAuthority(true, 2, false, true, true, false), candidate("software_info"));
        checks += unsafe(new SnapshotAuthority(true, 2, true, false, true, false), candidate("software_info"));
        checks += unsafe(new SnapshotAuthority(true, 2, true, true, false, false), candidate("software_info"));
        checks += unsafe(SAFE, candidate("software_info", Confidence.MEDIUM, true, true, true, true));
        checks += unsafe(SAFE, candidate("software_info", Confidence.HIGH, true, false, true, true));
        checks += unsafe(SAFE, candidate("software_info", Confidence.HIGH, true, true, false, true));
        checks += unsafe(SAFE, candidate("software_info", Confidence.HIGH, true, true, true, false));
        checks += unsafe(SAFE, candidate("software_info", Confidence.HIGH, false, true, true, true));
        assertEquals(43, checks);
    }

    private int unsafe(SnapshotAuthority authority, Candidate candidate) {
        return expect(Action.FALLBACK_GUIDE,
                base().screen(Screen.ABOUT_PHONE).authority(authority).candidate(candidate).build());
    }

    private int expect(Action expected, Observation observation) {
        assertEquals(expected, planner.decide(observation).action());
        return 1;
    }

    private AutopilotContract.Builder base() {
        return Observation.builder().authority(SAFE);
    }

    private Candidate candidate(String key) {
        return candidate(key, Confidence.HIGH, true, true, true, true);
    }

    private Candidate candidate(
            String key, Confidence confidence, boolean clickable,
            boolean unique, boolean margin, boolean fresh) {
        return new Candidate(key, confidence, clickable, unique, margin, fresh);
    }
}

File: docs/artifacts/remote-install-autopilot/reference/AdaptiveAutopilotPlannerTest.kt
package reference.autopilot

private val SAFE = SnapshotAuthority(true, 2, true, true, true)
private fun c(key: String, confidence: Confidence = Confidence.HIGH, clickable: Boolean = true, unique: Boolean = true, margin: Boolean = true, fresh: Boolean = true) = Candidate(key, confidence, clickable, unique, margin, fresh)
private fun o(vararg x: Pair<String, Any?>): Observation {
    val m = x.toMap()
    @Suppress("UNCHECKED_CAST")
    return Observation(
        androidApi = m["androidApi"] as? Int ?: 34,
        oem = m["oem"] as? String ?: "Samsung",
        accessibilityEnabled = m["accessibilityEnabled"] as? Boolean ?: true,
        adbConnected = m["adbConnected"] as? Boolean ?: false,
        supportConnected = m["supportConnected"] as? Boolean ?: false,
        restrictedSettingsRequired = m["restrictedSettingsRequired"] as? Boolean ?: false,
        wifiReady = m["wifiReady"] as? Boolean ?: true,
        wirelessPolicyBlocked = m["wirelessPolicyBlocked"] as? Boolean ?: false,
        previousPairingKnown = m["previousPairingKnown"] as? Boolean ?: false,
        reconnectAttempted = m["reconnectAttempted"] as? Boolean ?: false,
        screen = m["screen"] as? Screen ?: Screen.UNKNOWN,
        authority = if (m.containsKey("authority")) m["authority"] as SnapshotAuthority? else SAFE,
        candidate = m["candidate"] as Candidate?,
        wirelessEnabled = m["wirelessEnabled"] as Boolean?,
        buildTapsDone = m["buildTapsDone"] as? Int ?: 0,
        pairCodeCandidates = m["pairCodeCandidates"] as? List<String> ?: emptyList(),
        pairingContextHigh = m["pairingContextHigh"] as? Boolean ?: false,
        requestActive = m["requestActive"] as? Boolean ?: true,
        directDevProbeAttempted = m["directDevProbeAttempted"] as? Boolean ?: false,
        directDevScreenRecognized = m["directDevScreenRecognized"] as? Boolean ?: false,
    )
}

fun main() {
    val e = AdaptiveAutopilotPlanner()
    var n = 0
    fun expect(action: Action, obs: Observation) { check(e.decide(obs).action == action) { "expected $action got ${e.decide(obs)}" }; n++ }
    expect(Action.DONE, o("supportConnected" to true))
    expect(Action.CONNECT_SUPPORT, o("adbConnected" to true))
    expect(Action.UNSUPPORTED_ANDROID, o("androidApi" to 29))
    expect(Action.ASK_ENABLE_ACCESSIBILITY, o("accessibilityEnabled" to false))
    expect(Action.OPEN_DEVELOPER_SETTINGS, o("screen" to Screen.APP))
    expect(Action.OPEN_DEVICE_INFO_SETTINGS, o("screen" to Screen.SETTINGS_HOME, "directDevProbeAttempted" to true))
    expect(Action.FALLBACK_GUIDE, o("oem" to "Xiaomi", "screen" to Screen.SETTINGS_HOME, "directDevProbeAttempted" to true))
    expect(Action.CLICK_SOFTWARE_INFO, o("screen" to Screen.ABOUT_PHONE, "candidate" to c("software_info")))
    for (i in 0..6) expect(Action.CLICK_BUILD_NUMBER, o("screen" to Screen.SOFTWARE_INFO, "candidate" to c("build_number"), "buildTapsDone" to i))
    expect(Action.OPEN_DEVELOPER_SETTINGS, o("screen" to Screen.SOFTWARE_INFO, "candidate" to c("build_number"), "buildTapsDone" to 7))
    expect(Action.WAIT_USER_CREDENTIAL, o("screen" to Screen.CREDENTIAL_PROMPT))
    expect(Action.CLICK_WIRELESS_DEBUGGING, o("screen" to Screen.DEVELOPER_OPTIONS, "candidate" to c("wireless_debugging")))
    expect(Action.ENABLE_WIRELESS_DEBUGGING, o("screen" to Screen.WIRELESS_DEBUGGING, "candidate" to c("wireless_debugging_toggle"), "wirelessEnabled" to false))
    expect(Action.CLICK_PAIR_WITH_CODE, o("screen" to Screen.WIRELESS_DEBUGGING, "candidate" to c("pair_with_code"), "wirelessEnabled" to true))
    expect(Action.ACCEPT_NETWORK_CONFIRMATION, o("screen" to Screen.NETWORK_CONFIRMATION, "candidate" to c("network_confirm_positive")))
    expect(Action.AUTO_PAIR_WITH_CODE, o("screen" to Screen.PAIRING_DIALOG, "pairCodeCandidates" to listOf("123456"), "pairingContextHigh" to true))
    expect(Action.SHOW_MANUAL_PAIR_CODE, o("screen" to Screen.PAIRING_DIALOG, "pairCodeCandidates" to listOf("123456","654321"), "pairingContextHigh" to true))
    expect(Action.SHOW_MANUAL_PAIR_CODE, o("screen" to Screen.PAIRING_DIALOG, "pairCodeCandidates" to listOf("123456"), "pairingContextHigh" to false))
    expect(Action.SHOW_MANUAL_PAIR_CODE, o("screen" to Screen.PAIRING_DIALOG, "pairCodeCandidates" to listOf("123456"), "pairingContextHigh" to true, "requestActive" to false))
    expect(Action.FALLBACK_GUIDE, o("screen" to Screen.ABOUT_PHONE, "authority" to SnapshotAuthority(true,1,true,true,true), "candidate" to c("software_info")))
    expect(Action.FALLBACK_GUIDE, o("screen" to Screen.ABOUT_PHONE, "authority" to SnapshotAuthority(false,2,true,true,true), "candidate" to c("software_info")))
    expect(Action.FALLBACK_GUIDE, o("screen" to Screen.ABOUT_PHONE, "authority" to SnapshotAuthority(true,2,true,true,true,true), "candidate" to c("software_info")))
    expect(Action.FALLBACK_GUIDE, o("screen" to Screen.ABOUT_PHONE, "authority" to SnapshotAuthority(true,2,false,true,true), "candidate" to c("software_info")))
    expect(Action.FALLBACK_GUIDE, o("screen" to Screen.ABOUT_PHONE, "candidate" to c("software_info", Confidence.MEDIUM)))
    expect(Action.FALLBACK_GUIDE, o("screen" to Screen.ABOUT_PHONE, "candidate" to c("software_info", unique=false)))
    expect(Action.FALLBACK_GUIDE, o("screen" to Screen.ABOUT_PHONE, "candidate" to c("software_info", margin=false)))
    expect(Action.FALLBACK_GUIDE, o("screen" to Screen.ABOUT_PHONE, "candidate" to c("software_info", fresh=false)))
    expect(Action.WAIT_STABLE, o("screen" to Screen.WIRELESS_DEBUGGING, "candidate" to c("pair_with_code")))
    expect(Action.WAIT_STABLE, o("screen" to Screen.PAIRING_DIALOG, "authority" to SnapshotAuthority(true,1,true,true,true), "pairCodeCandidates" to listOf("123456"), "pairingContextHigh" to true))
    expect(Action.ASK_ALLOW_RESTRICTED_SETTINGS, o("accessibilityEnabled" to false, "restrictedSettingsRequired" to true))
    expect(Action.ASK_CONNECT_WIFI, o("wifiReady" to false))
    expect(Action.POLICY_BLOCKED, o("wirelessPolicyBlocked" to true))
    expect(Action.TRY_ADB_RECONNECT, o("previousPairingKnown" to true, "reconnectAttempted" to false))
    expect(Action.OPEN_DEVELOPER_SETTINGS, o("previousPairingKnown" to true, "reconnectAttempted" to true, "screen" to Screen.APP))
    println("PASS $n Kotlin reference checks")
}
